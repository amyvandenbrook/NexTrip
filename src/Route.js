import React from 'react'

export const ChooseRoute = ( props ) => {
    if (props.routeLoaded) {
        return (
            <React.Fragment>
                <p>Please select a route from the options below to see the directions and stops available for you!</p>
                <select id="Route" data-testid="Select-Route" name="Route" onChange={props.routeHandler}>
                    {!props.routeChanged ? <option value={-1}>Select a Route</option> : null};
                    {props.routeData.map(data => (
                        <option key={data.Route} value={data.Route}>{data.Description}</option>
                    ))}
                </select>
                <i className="fas fa-arrow-circle-down"></i>
            </React.Fragment>
        )
    }
    else {
        return <div>Loading</div>
    }
}