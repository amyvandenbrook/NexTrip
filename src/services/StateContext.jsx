import React, { createContext, useReducer } from 'react'
import Reducer from './Reducer.jsx'

const StateContext = createContext()

const StateProvider = ({ children, initialState }) => {
    const defaultState = {
        routeData: [],
        route: undefined,
        routeChanged: false,
        directionData: [],
        direction: '',
        stopData: [],
    }
    const [state, dispatch] = useReducer(Reducer, {
        ...defaultState,
        ...initialState,
    })

    return (
        <StateContext.Provider
            value={{
                ...state,
                dispatch,
            }}
        >
            {children}
        </StateContext.Provider>
    )
}

export { StateContext, StateProvider }
