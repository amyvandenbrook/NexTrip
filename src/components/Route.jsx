import React, { useContext, useEffect } from 'react'
import { StateContext } from '../services/StateContext.jsx'
import { getRoutes } from '../services/AxiosService.jsx'

const Route = () => {
    const { routeChanged, routeData, dispatch } = useContext(StateContext)

    useEffect(() => {
        getRoutes().then((result) => {
            dispatch({ type: 'set_route_data', payload: result.data })
        })
    }, [dispatch])

    const setRoute = (event) => {
        dispatch({ type: 'set_route', payload: event.target.value })
    }

    return (
        <React.Fragment>
            <p>
                Please select a route from the options below to see the
                directions and stops available for you!
            </p>
            <select
                id="Route"
                data-testid="Select-Route"
                name="Route"
                onChange={setRoute}
            >
                {!routeChanged ? (
                    <option value={-1}>Select a Route</option>
                ) : null}
                ;{' '}
                {routeData.map((data) => (
                    <option key={data.Route} value={data.Route}>
                        {data.Description}
                    </option>
                ))}
            </select>
            <i className="fas fa-arrow-circle-down"></i>
        </React.Fragment>
    )
}

export default Route
