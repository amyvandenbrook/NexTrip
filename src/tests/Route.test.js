import React from 'react'
import { act, create } from 'react-test-renderer'
import { render, fireEvent, screen } from '@testing-library/react'
import axios from 'axios'
import { StateProvider } from '../services/StateContext.jsx'
import Route from '../components/Route.jsx'

jest.mock('axios')
axios.get = jest.fn()

describe('Route', () => {
    let route
    let component

    test('is created properly before getting any routes', async () => {
        const mockReturnedData = ['testRoute']
        axios.get.mockImplementationOnce(() =>
            Promise.resolve({
                status: 200,
                data: mockReturnedData,
            })
        )
        await act(() => {
            component = create(
                <StateProvider>
                    <Route />
                </StateProvider>
            )
        })
        route = component.toJSON()
        expect(route).toMatchSnapshot()
    })

    test('changing selection updates routeChanged', () => {
        render(
            <StateProvider>
                <Route />
            </StateProvider>
        )
        const routeSelector = screen.getByTestId('Select-Route')

        fireEvent.change(routeSelector)

        expect(
            screen.getByTestId('Select-Route').children.length
        ).toStrictEqual(0)
    })
})
