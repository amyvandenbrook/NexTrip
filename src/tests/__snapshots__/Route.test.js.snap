// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route is created properly before getting any routes 1`] = `
Array [
  <p>
    Please select a route from the options below to see the directions and stops available for you!
  </p>,
  <select
    className="route_selector"
    data-testid="Select-Route"
    id="Route"
    name="Route"
    onChange={[Function]}
  >
    <option
      value={-1}
    >
      Select a Route
    </option>
    ;
     
    <option />
  </select>,
  <i
    className="fas fa-arrow-circle-down"
  />,
]
`;
